opm/common/utility/CSRGraphFromCoordinates.hpp

```diff
@@ -0,0 +1,490 @@
+            BaseVertexID numCols_{};
+
+            // ---------------------------------------------------------
+            // Implementation of read()/write()
+            // ---------------------------------------------------------
+
+            template <typename T, class A, class MessageBufferType>
+            void writeVector(const std::vector<T,A>& vec,
+                             MessageBufferType&      buffer) const
+            {
+                const auto n = vec.size();
+                buffer.write(n);
+
+                for (const auto& x : vec) {
+                    buffer.write(x);
+                }
+            }
+
+            template <class MessageBufferType, typename T, class A>
+            void readVector(MessageBufferType& buffer,
+                            std::vector<T,A>&  vec)
+            {
+                auto n = 0 * vec.size();
+                buffer.read(n);
+
+                vec.resize(n);
+
+                for (auto& x : vec) {
+                    buffer.read(x);
+                }
+            }
+
+            // ---------------------------------------------------------
+            // Implementation of merge()
+            // ---------------------------------------------------------
+
+            /// Incorporate new, coordinate format contributions into
+            /// existing, possibly empty, CSR mapping structure.
+            ///
+            /// On exit the ia_ array holds the proper start pointers while
+            /// ja_ holds the corresponding column indices albeit possibly
+            /// repeated and unsorted.
+            ///
+            /// \param[in] rows Row indices of all, possibly repeated,
+            ///    coordinate format input contributions.  Start pointers \c
+            ///    ia_ updated to account for new entries.
+            ///
+            /// \param[in] cols Column index of coordinate format intput
+            ///    structure.  Inserted into \c ja_ according to its
+            ///    corresponding row index.
+            ///
+            /// \param[in] maxRowID Maximum index in \p rows.  Needed to
+            ///    ensure proper size of \c ia_.
+            ///
+            /// \param[in] maxColID Maximum index in \p cols.
+            void assemble(const Neighbours& rows,
+                          const Neighbours& cols,
+                          BaseVertexID      maxRowID,
+                          BaseVertexID      maxColID);
+
+            /// Sort column indices per row and compress repeated column
+            /// indices down to a single unique element per row.  Sum
+            /// repeated values
+            ///
+            /// On exit the \c ia_ and \c ja_ arrays all have their
+            /// expected, canonical structure.
+            ///
+            /// \param[in] maxNumVertices Maximum number of vertices
+            ///    supported by final compressed mapping structure.  Ignored
```

Comment:
I assume that this is intended to support building a graph with a certain number of nodes, even if some of those (at highest indices) are not connected. We should throw instead of ignore if the user requests a too small number here, as the user will expect numRows() equal to maxNumVertices after the merge() operation.

Reply:
> We should throw instead of ignore if the user requests a too small number here, as the user will expect `numRows()` equal to `maxNumVertices` after the `merge()` operation.

There is already a check in place [ahead of the only `compress()`](https://github.com/OPM/opm-common/blob/3031759c82a4e8137e763d2f3a08bc805d1b2e94/opm/common/utility/CSRGraphFromCoordinates_impl.hpp#L153-L162) call, but I've pushed an update to add that check within `CSR::compress()` as well.